### PR TITLE
EZP-30057: Implemented Schema Exporter, Importer and Builder

### PR DESCRIPTION
**Pull request by alongosz**
Implements [EZP-30057](https://jira.ez.no/browse/EZP-30057), needed for [EZP-29938: PostgreSQL support](https://jira.ez.no/browse/EZP-29938).

This PR provides following features available via APIs:
1. `\EzSystems\DoctrineSchema\API\SchemaImporter` which imports given Yaml source string or Yaml file into `\Doctrine\DBAL\Schema` object.
2. `\EzSystems\DoctrineSchema\API\SchemaExporter` which exports given `\Doctrine\DBAL\Schema` object to the custom Yaml format.
3. `\EzSystems\DoctrineSchema\API\Builder\SchemaBuilder`

`SchemaBuilder` is event-driven. To hook into the process of building schema, a custom `EventSubscriber` is required, e.g.
```php
use EzSystems\DoctrineSchema\API\Event\SchemaBuilderEvent;
use EzSystems\DoctrineSchema\API\Event\SchemaBuilderEvents;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;

class BuildSchemaSubscriber implements EventSubscriberInterface
{
    /**
     * @var string
     */
    private $schemaFilePath;

    public function __construct(string $schemaFilePath)
    {
        $this->schemaFilePath = $schemaFilePath;
    }

    /**
     * Returns an array of events this subscriber wants to listen to.
     *
     * @return string[]
     */
    public static function getSubscribedEvents()
    {
        return [
            SchemaBuilderEvents::BUILD_SCHEMA => ['onBuildSchema', 200],
        ];
    }

    /**
     * @param \EzSystems\DoctrineSchema\API\Builder\SchemaBuilderEvent $event
     */
    public function onBuildSchema(SchemaBuilderEvent $event)
    {
        $event
            ->getSchemaBuilder()
            ->importSchemaFromFile($this->schemaFilePath);
    }
}
```
Schema provided in this way can be imported into Schema object by e.g.:
```php
    public function __construct(SchemaBuilder $schemaBuilder)
    {
        $this->schemaBuilder = $schemaBuilder;
    }

    public function importSchema()
    {
        $schema = $this->schemaBuilder->buildSchema();
        // ...
    }
```

All the described services are autowired, so injecting them into custom Symfony services is quite easy.

The documentation of the custom Yaml format will be provided shortly.

**TODO**:
- [x] Implement Schema Importer
- [x] Implement Schema Builder
- [x] Provide more detailed description for exposed APIs

### CODE REVIEW

File: src/lib/Database/DbPlatformFactory.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace EzSystems\DoctrineSchema\Database;
+
+use Doctrine\DBAL\Platforms\AbstractPlatform;
+use EzSystems\DoctrineSchema\API\DbPlatformFactory as APIDbPlatformFactory;
+
+class DbPlatformFactory implements APIDbPlatformFactory
+{
+    /**
+     * @var array
```
Comment (adamwojs) [inline src/lib/Database/DbPlatformFactory.php]:
```suggestion
     * @var \EzSystems\DoctrineSchema\Database\DbPlatform\DbPlatform[]
```

File: src/lib/API/Builder/SchemaBuilder.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+     * @return \Doctrine\DBAL\Schema\Schema
+     */
+    public function buildSchema(): Schema;
+
+    /**
+     * Import Schema from Yaml schema definition file into Schema object.
+     *
+     * @param string $schemaFilePath
+     *
+     * @return mixed
```
Comment (mikadamczyk) [inline src/lib/API/Builder/SchemaBuilder.php]:
```suggestion
     * @return \Doctrine\DBAL\Schema\Schema
```
?

File: src/lib/API/Event/SchemaBuilderEvent.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     * @var \EzSystems\DoctrineSchema\API\Builder\SchemaBuilder
+     */
+    private $schemaBuilder;
+
+    /**
+     * @var \Doctrine\DBAL\Schema\Schema
+     */
+    private $schema;
+
+    public function __construct(SchemaBuilder $schemaBuilder, Schema $schema)
```
Comment (mikadamczyk) [inline src/lib/API/Event/SchemaBuilderEvent.php]: Are we using docblocks?

File: src/lib/Database/DbPlatform/PostgreSqlDbPlatform.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    /**
+     * Returns the SQL snippet to drop an existing table.
+     *
+     * @param \Doctrine\DBAL\Schema\Table|string $table
+     *
+     * @return string
+     *
+     * @throws \InvalidArgumentException
+     */
+    public function getDropTableSQL($table)
```
Comment (mikadamczyk) [inline src/lib/Database/DbPlatform/PostgreSqlDbPlatform.php]:
```suggestion
    public function getDropTableSQL($table): string
```

File: src/lib/Database/DbPlatform/SqliteDbPlatform.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        return true;
+    }
+
+    /**
+     * Override default behavior of Sqlite db platform not to throw exception for unsupported operation of dropping FKs.
+     *
+     * {@inheritdoc}
+     */
+    public function getDropForeignKeySQL($foreignKey, $table)
```
Comment (mikadamczyk) [inline src/lib/Database/DbPlatform/SqliteDbPlatform.php]:
```suggestion
    public function getDropForeignKeySQL($foreignKey, $table): string
```

File: src/lib/Database/DbPlatform/SqliteDbPlatform.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+        return '-- ';
+    }
+
+    /**
+     * Override default behavior of Sqlite db platform not to throw exception for unsupported operation of creating FKs.
+     *
+     * {@inheritdoc}
+     */
+    public function getCreateForeignKeySQL(ForeignKeyConstraint $foreignKey, $table)
```
Comment (mikadamczyk) [inline src/lib/Database/DbPlatform/SqliteDbPlatform.php]:
```suggestion
    public function getCreateForeignKeySQL(ForeignKeyConstraint $foreignKey, $table): string
```

File: src/lib/API/Builder/SchemaBuilder.php (hunk 2 above)
Comment (alongosz) [reply src/lib/API/Builder/SchemaBuilder.php]: :scream:

**Pull request merged**